```{r}
library(SNPRelate)
library(ggplot2)

# Convert bsnp PLINK files to GDS
snpgdsBED2GDS(bed.fn = "bsnp_filt_plink.vcf.gz.bed",
              bim.fn = "bsnp_filt_plink.vcf.gz.bim",
              fam.fn = "bsnp_filt_plink.vcf.gz.fam",
              out.gdsfn = "bsnp_filt_plink.vcf.gz.gds")
```

```{r}
#open the .gds file
b_genofile <- snpgdsOpen("bsnp_filt_plink.vcf.gz.gds")

#cluster by PCA
b_pca <- snpgdsPCA(b_genofile, autosome.only = FALSE)  # autosome.only=FALSE if you want all chromosomes
```

```{r}
#get summary for the gds file
snpgdsSummary(b_genofile)
```

```{r}
# Sample IDs
sample.id <- read.gdsn(index.gdsn(b_genofile, "sample.id"))

# PC scores
#get the percent of variance explained by each PC
b_pc.percent <- b_pca$varprop * 100
b_pc.df <- data.frame(PC1 = b_pca$eigenvect[,1], 
                    PC2 = b_pca$eigenvect[,2], 
                    Sample = sample.id)

#name your population groups   
b_pc.df$Population <- factor(c("pop1", "pop2",))

#plot PCs
ggplot(b_pc.df, aes(x = PC1, y = PC2, label = Sample,color = Population)) +
  geom_point() +
  labs(x = paste0("PC1 (", round(b_pc.percent[1], 2), "%)"),
       y = paste0("PC2 (", round(b_pc.percent[2], 2), "%)"), title = "Populations (Bis-SNP) (Filtered)") +
  theme_classic()
```
```{r}
#close the gds file
snpgdsClose(b_genofile) 
```

```{r}

# Convert cgmap PLINK files to GDS
snpgdsBED2GDS(bed.fn = "cgmap_filt_plink.vcf.gz.bed",
              bim.fn = "cgmap_filt_plink.vcf.gz.bim",
              fam.fn = "cgmap_filt_plink.vcf.gz.fam",
              out.gdsfn = "cgmap_filt_plink.vcf.gz.gds")
```

```{r}
#open the .gds file
c_genofile <- snpgdsOpen("cgmap_filt_plink.vcf.gz.gds")

#cluster by PCA
c_pca <- snpgdsPCA(c_genofile, autosome.only = FALSE)  # autosome.only=FALSE if you want all chromosomes
```

```{r}
#get summary fo the gds file
#snpgdsClose(c_genofile) 
```


```{r}
# Sample IDs
sample.id <- read.gdsn(index.gdsn(c_genofile, "sample.id"))

# PC scores
#get the percent of variance explained by each PC
c_pc.percent <- c_pca$varprop * 100
c_pc.df <- data.frame(PC1 = c_pca$eigenvect[,1], 
                    PC2 = c_pca$eigenvect[,2], 
                    Sample = sample.id)
#name your population groups 
c_pc.df$Group <- factor(c("pop1", "pop2"))

#plot PCs
ggplot(c_pc.df, aes(x = PC1, y = PC2, label = Sample,color = Group)) +
  geom_point() +
  labs(x = paste0("PC1 (", round(c_pc.percent[1], 2), "%)"),
       y = paste0("PC2 (", round(c_pc.percent[2], 2), "%)"), title = "Populations (CGmapTools) (Filtered)") +
  theme_classic()



```



```{r}
snpgdsSummary(c_genofile)
```